### ISSUE TITLE
Fix package has-a spec relationship

### ISSUE THREAD

**Issue by mathstuf** [LLNL/spack#5996]
A package should not contain a (concretized) spec. A spec may be concretized, but I don't really know what it means for a package to be concretized.

See #5600 for some related discussion.

Cc: @tgamblin @scheibelp

**Comment by tgamblin** [comment]
I guess I'm confused.  There are three things at the core of Spack:

1. Package classes, which contain metadata and constraints for resolving dependencies and package configuration.
2. Spec instances, which represent package configurations, i.e. the dependency DAG.
3. Package instances, which are basically installers for concrete `Spec`s.

Specs are dynamic; Packages do two things: the class has the static constraints (invariants) and the instance is the installer.  They're conflated like that b/c this has been shown (in `Homebrew`, at least) to make for a nice package DSL.  I tend to think that's true.

There are currently some issues with the lifecycles of those things.  What I see as the problems are:

1. `Package.__init__()` does some setup that should really be in the metaclass (i.e. associated with the class, not the instance).  I'd have to review the code but I think this is entirely vestigial -- maybe it was expedient when I first wrote the class, but now it's hacky and in the way.
2. As a result of (1), most of Spack, when it really needs the package *class*, calls `Repo.get()` with just the *name* of the package, to get an instance of `Package` that it uses only for metadata.  This stuff should be getting the class by name instead.
2. `Repo` currently caches package instances by `Spec`, and those `Package` instances have a `Spec`.  This is unnecessary and causes confusion.  I think I thought that we'd need to cache instances after loading package classes, but really just caching the classes is fine.

I think the lifecycle should look like this:

1. User (or something else) provides an abstract Spec.
2. Concretize the spec.
3. Get the class for the root, and use that to construct a `Package` instance for *just this install*.
4. Install using the package instance.  *While this is happening*, the `Package` has a `self.spec` (since the methods in `Package` need to look at the `Spec`).
5. After that, the package forgets about the `Spec` until someone calls install on it again.

RE: #5600 Extensions:  The `Package.extends()` method should really be on `Spec`, since whether or not something "extends" something else depends on how it was installed.  i.e., some packages only extend Python when build with an optional variant.  In general, if it is a dynamic thing, it belongs in `Spec`.  I think we added `extends()` before there were optional dependencies, and it's remained somewhat awkward since.

Does that help?

**Comment by mathstuf** [comment]
> Install using the package instance. While this is happening, the Package has a self.spec (since the methods in Package need to look at the Spec).

Why does it have to be `self.spec` and not just an argument to the install method?

**Comment by tgamblin** [comment]
There's not *just* install.  The install process consists of a lot more than install.  There are fetch, stage, patch, install, setup_dependent_environment, etc. methods that also need the `Spec`.

Why do you *not* want this to be an instance variable on the installer object?

**Comment by mathstuf** [comment]
> Why do you not want this to be an instance variable on the installer object?

Because it only has meaning during an actual install: it should only be accessible/used during that process.

Maybe the class which performs the install should have a `Package` and a concrete `Spec`?

**Comment by tgamblin** [comment]
> Maybe the class which performs the install should have a Package and a concrete Spec?

I think I'm proposing something similar: the package should *only* be instantiated before install, and only live until the end of an install.  Repo should return Package class objects, not instances.

So effectively that's the same thing, right?  The other option would be to have a separate `PackageInstaller` class, but it would still need to call into instance methods on the `Package`, and I think `Package` would still need to have a `spec` attribute because all the code in `package.py` subclasses `Package` and needs to talk about the `Spec`.

So basically I think the right initial refactor would be to de-conflate the class-scope stuff from the instance methods in `Package`, then think about adding a separate installer object if we need it.

Sound good?

**Comment by mathstuf** [comment]
Yeah, that sounds sensible. It could also be something related to `Spec` directly; it already has a `self.package` attribute (which makes more sense).
